File: bindings/java-jni/src/main/java/org/maplibre/nativejni/runtime/OfflineOperationKind.java
package org.maplibre.nativejni.runtime;

/** Offline database operation kind reported by completion events. */
public final class OfflineOperationKind {
  public static final OfflineOperationKind AMBIENT_CACHE =
      new OfflineOperationKind(1, "AMBIENT_CACHE");
  public static final OfflineOperationKind REGION_CREATE =
      new OfflineOperationKind(2, "REGION_CREATE");
  public static final OfflineOperationKind REGION_GET = new OfflineOperationKind(3, "REGION_GET");
  public static final OfflineOperationKind REGIONS_LIST =
      new OfflineOperationKind(4, "REGIONS_LIST");
  public static final OfflineOperationKind REGIONS_MERGE_DATABASE =
      new OfflineOperationKind(5, "REGIONS_MERGE_DATABASE");
  public static final OfflineOperationKind REGION_UPDATE_METADATA =
      new OfflineOperationKind(6, "REGION_UPDATE_METADATA");
  public static final OfflineOperationKind REGION_GET_STATUS =
      new OfflineOperationKind(7, "REGION_GET_STATUS");
  public static final OfflineOperationKind REGION_SET_OBSERVED =
      new OfflineOperationKind(8, "REGION_SET_OBSERVED");
  public static final OfflineOperationKind REGION_SET_DOWNLOAD_STATE =
      new OfflineOperationKind(9, "REGION_SET_DOWNLOAD_STATE");
  public static final OfflineOperationKind REGION_INVALIDATE =
      new OfflineOperationKind(10, "REGION_INVALIDATE");
  public static final OfflineOperationKind REGION_DELETE =
      new OfflineOperationKind(11, "REGION_DELETE");

  private final int nativeValue;
  private final String name;

  private OfflineOperationKind(int nativeValue, String name) {
    this.nativeValue = nativeValue;
    this.name = name;
  }

  public int nativeValue() {
    return nativeValue;
  }

  public int rawValue() {
    return nativeValue;
  }

  public static OfflineOperationKind fromNative(int nativeValue) {
    return switch (nativeValue) {
      case 1 -> AMBIENT_CACHE;
      case 2 -> REGION_CREATE;
      case 3 -> REGION_GET;
      case 4 -> REGIONS_LIST;
      case 5 -> REGIONS_MERGE_DATABASE;
      case 6 -> REGION_UPDATE_METADATA;
      case 7 -> REGION_GET_STATUS;
      case 8 -> REGION_SET_OBSERVED;
      case 9 -> REGION_SET_DOWNLOAD_STATE;
      case 10 -> REGION_INVALIDATE;
      case 11 -> REGION_DELETE;
      default -> new OfflineOperationKind(nativeValue, null);
    };
  }

  @Override
  public boolean equals(Object other) {
    return other instanceof OfflineOperationKind value && nativeValue == value.nativeValue;
  }

  @Override
  public int hashCode() {
    return Integer.hashCode(nativeValue);
  }

  @Override
  public String toString() {
    return name != null ? name : "OfflineOperationKind(" + nativeValue + ")";
  }
}

File: bindings/kotlin/src/commonMain/kotlin/org/maplibre/nativeffi/runtime/OfflineOperationKind.kt
package org.maplibre.nativeffi.runtime

import kotlin.jvm.JvmInline

/** Offline operation kind reported by completion events. */
@JvmInline
public value class OfflineOperationKind(public val nativeValue: Int) {
  public companion object {
    public val AMBIENT_CACHE: OfflineOperationKind = OfflineOperationKind(1)
    public val REGION_CREATE: OfflineOperationKind = OfflineOperationKind(2)
    public val REGION_GET: OfflineOperationKind = OfflineOperationKind(3)
    public val REGIONS_LIST: OfflineOperationKind = OfflineOperationKind(4)
    public val REGIONS_MERGE_DATABASE: OfflineOperationKind = OfflineOperationKind(5)
    public val REGION_UPDATE_METADATA: OfflineOperationKind = OfflineOperationKind(6)
    public val REGION_GET_STATUS: OfflineOperationKind = OfflineOperationKind(7)
    public val REGION_SET_OBSERVED: OfflineOperationKind = OfflineOperationKind(8)
    public val REGION_SET_DOWNLOAD_STATE: OfflineOperationKind = OfflineOperationKind(9)
    public val REGION_INVALIDATE: OfflineOperationKind = OfflineOperationKind(10)
    public val REGION_DELETE: OfflineOperationKind = OfflineOperationKind(11)

    internal fun fromNative(nativeValue: UInt): OfflineOperationKind =
      fromNative(nativeValue.toInt())

    internal fun fromNative(nativeValue: Int): OfflineOperationKind =
      OfflineOperationKind(nativeValue)
  }
}
